keps/sig-auth/3221-multiple-authorization-webhooks/README.md

```diff
@@ -0,0 +1,479 @@
+
+## Motivation
+
+Cluster administrators should be able to specify more than one authorization webhook in the API Server handler chain. They also need to be able to declaratively configure the authorizer chain using a configuration file.
+
+It should also be easy to say when to Deny requests, for example, when a webhook is unreachable.
+
+### Goals
+
+- Define a configuration file format for configuring Kubernetes API Server Authorization chain.
+- Allow ordered definition of authorization modes.
+- Allow definition of multiple webhooks in the authorization chain.
+- Enable user to define the policy when a webhook can't be reached due to network errors etc.
+- Enable more defnitive results for a `SubjectAccessReview` by allowing config to distinguish webhooks which might deny access.
+
+### Non-Goals
+
+- Allow resource/user based pre-filtering of webhooks to prevent unnecessary invocations.
+- Modify the webhook authentication format.
+- Hot reload of the configuration.
+
+## Proposal
+
+Add a configuration format having specific precedence order and defined failure modes for configuring authorizer chain.
+
+```yaml
+apiVersion: apiserver.config.k8s.io/v1
+kind: AuthorizationConfiguration
+authorizers:
+- type: Webhook
+  webhook:
+    kubeConfigFile: <path-to-deny-webhook-config>
+    authorizedTTL: 5m
+    unauthorizedTTL: 30s
+    subjectAccessReviewVersion: v1
+    onError: Deny
+- type: Node
+- type: RBAC
+- type: Webhook
+  webhook:
+    kubeConfigFile: <path-to-allow-webhook-config>
+    authorizedTTL: 5m
+    unauthorizedTTL: 30s
+    subjectAccessReviewVersion: v1
+    onError: NoOpinion
+- type: ABAC
+  abac:
+    policyFile: <path-to-policy-file>
+```
+
+### User Stories (Optional)
+
+#### Story 1: Protecting installed CRDs
+
+Having certain Custom Resource Definitions available at cluster startup has been a long requested feature. One of the blockers for having a controller reconciling those CRDs is to have a protection mechanism for them, which can be achieved through Authorization Webhooks. However since `kube-apiserver` only allows specifying a single webhook, this would result in cluster administrators being not able to specify their own. Support for multiple webhooks would make this possible.
+
+Relevant Discussion Thread is sig-apimachinery: https://groups.google.com/g/kubernetes-sig-api-machinery/c/MBa19WTETMQ
+
+#### Story 2: Preventing unnecessarily nested webhooks
+
+Consider a system administrator who would like to apply a set of validations to certain requests before handing it off to webhooks defined using frameworks like Open Policy Agent.
+
+They would have to nested webhooks within the one added to the auth chain to have the intended effect. This enhancement allows the administrator to configure this behaviour via a structured API and invoke the additional webhook only when relevant.
```

Comment:
Can you expand on why a nested webhook could not also have logic to avoid unnecessary calls?

Reply:
Having a single webhook with nested webhooks introduces a single point of failure. 
A way to supply multiple authorization webhooks with defined `onError` behaviours would be a more robust way of achieving this with more predicatble decisions  natively within Kubernetes.